File: 2241/main.go
package main

import "fmt"

type ATM struct {
	value []int64
	cnt   []int64
}

func Constructor() ATM {
	return ATM{cnt: make([]int64, 5), value: []int64{20, 50, 100, 200, 500}}
}

func (atm *ATM) Deposit(banknotesCount []int) {
	for i := range banknotesCount {
		atm.cnt[i] += int64(banknotesCount[i])
	}
}

func (atm *ATM) Withdraw(amount int) []int {
	res := make([]int, 5)
	for i := 4; i >= 0; i-- {
		res[i] = int(min(atm.cnt[i], int64(amount)/atm.value[i]))
		amount -= res[i] * int(atm.value[i])
	}
	if amount > 0 {
		return []int{-1}
	}
	for i := 0; i < 5; i++ {
		atm.cnt[i] -= int64(res[i])
	}
	return res
}

func main() {
	obj := Constructor()
	obj.Deposit([]int{0,0,1,2,1})
	fmt.Println(obj.Withdraw(600))
	obj.Deposit([]int{0,1,0,1,1})
	fmt.Println(obj.Withdraw(600))
}

File: 2241/main.rs

struct ATM {
    cnt: Vec<i64>,
    value: Vec<i64>,
}

impl ATM {
    fn new() -> Self {
        ATM {
            cnt: vec![0,0,0,0,0],
            value: vec![20, 50, 100, 200, 500],
        }
    }

    fn deposit(&mut self, banknotes_count: Vec<i32>) {
        for i in 0..5 {
            self.cnt[i] += banknotes_count[i] as i64;
        }
    }

    fn withdraw(&mut self, mut amount: i32) -> Vec<i32> {
        let mut res = vec![0; 5];
        for i in (0..5).rev() {
            res[i] = std::cmp::min(self.cnt[i], amount as i64 / self.value[i]) as i32;
            amount -= res[i] * self.value[i] as i32;
        }
        if amount > 0 {
            vec![-1]
        } else {
            for i in 0..5 {
                self.cnt[i] -= res[i] as i64;
            }
            res
        }
    }
}

fn main() {
    let mut obj = ATM::new();
    obj.deposit(vec![0,0,1,2,1]);
    println!("{:#?}", obj.withdraw(600));
}